openlr/paths_connector.cpp

```diff
@@ -0,0 +1,236 @@
+      return false;
+    }
+  }
+
+  ASSERT_EQUAL(resultPath.size(), points.size() - 1, ());
+
+  return true;
+}
+
+bool PathsConnector::FindShortestPath(Graph::Edge const & from, Graph::Edge const & to,
+                                      FunctionalRoadClass const frc, uint32_t const maxPathLength,
+                                      Graph::EdgeVector & path)
+{
+  // TODO(mgsergio): Turn Dijkstra to A*.
+
+  uint32_t const kLengthToleranceM = 10;
+
+  struct State
+  {
+    State(Graph::Edge const & e, uint32_t const s) : m_edge(e), m_score(s) {}
+
+    bool operator>(State const & o) const
+    {
+      return make_tuple(m_score, m_edge) > make_tuple(o.m_score, o.m_edge);
+    }
+
+    Graph::Edge m_edge;
+    uint32_t m_score;
+  };
+
+  priority_queue<State, vector<State>, greater<State>> q;
+  map<Graph::Edge, uint32_t> scores;
+  map<Graph::Edge, Graph::Edge> links;
+
+  q.emplace(from, 0);
+  scores[from] = 0;
+
+  while (!q.empty())
+  {
+    auto const state = q.top();
+    q.pop();
+
+    auto const & u = state.m_edge;
+    // TODO(mgsergio): Unify names: use either score or distance.
+    auto const us = state.m_score;
+
+    if (us > maxPathLength + kLengthToleranceM)
+      continue;
+
+    if (us > scores[u])
+      continue;
+
+    if (u == to)
+    {
+      for (auto e = u; e != from; e = links[e])
+        path.push_back(e);
+      reverse(begin(path), end(path));
+      return true;
+    }
+
+    Graph::EdgeVector edges;
+    m_graph.GetOutgoingEdges(u.GetEndJunction(), edges);
+    for (auto const & e : edges)
+    {
+      // TODO(mgsergio): Use frc to filter edges.
+
+      auto const it = scores.find(e);
+      auto const eScore = us + EdgeLength(e);
+      if (it == end(scores) || it->second > eScore)
+      {
+        scores[e] = eScore;
+        links[e] = u;
+        q.emplace(e, eScore);
+      }
+    }
+  }
+
+  return false;
+}
+
+bool PathsConnector::ConnectAdjacentCandidateLines(Graph::EdgeVector const & from,
+                                                   Graph::EdgeVector const & to,
+                                                   FunctionalRoadClass const frc,
+                                                   uint32_t const distanceToNextPoint,
+                                                   Graph::EdgeVector & resultPath)
+
+{
+  ASSERT(!to.empty(), ());
+
+  if (auto const skip = PathOverlappingLen(from, to))
+  {
+    copy(begin(from), end(from), back_inserter(resultPath));
+    copy(begin(to) + skip, end(to), back_inserter(resultPath));
+    return true;
+  }
+
+  ASSERT(from.back() != to.front(), ());
+
+  Graph::EdgeVector shortestPath;
+  auto const found =
+      FindShortestPath(from.back(), to.front(), frc, distanceToNextPoint, shortestPath);
+  if (!found)
+    return false;
+
+  copy(begin(from), end(from), back_inserter(resultPath));
+  copy(begin(shortestPath), end(shortestPath), back_inserter(resultPath));
+  copy(next(begin(to)), end(to), back_inserter(resultPath));
```

Comment:
Please, add at least comment why do we need to call next.

Reply:
Done.